### ISSUE TITLE
NTI: Expression type determined incorrectly

### ISSUE THREAD

**Issue by crdev** [google/closure-compiler#650]
New type inference compilation for

```js
/** @constructor */
function Descriptor() {
    /**
     * @type {string|undefined}
     */
    this.className;
}

...
// In another constructor, where descriptor is of type Descriptor:
/** @type {?string} */
this._className = descriptor.className || null;
```
results in
```
WARNING - The right side in the assignment is not a subtype of the left side.
left side  : null|string
right side : null|string|undefined

    this._className = descriptor.className || null;
    ^
```

**Comment by Dominator008** [comment]
The repro case does not give a warning any more.
